[[ Bug 16120 ]] Hardlink iOS simulator copied files rather than symlink

livecodesebastien opened the pull request:
iOS Simulator 9.0 make iPhoneProxy hang if symlinks are used in the app bundle.
Using hardlinks instead fixes the issue, and the application can be started after deployment

livecodesebastien committed:
[[ Bug 16120 ]] Hardlink iOS simulator copied files rather than symlink

iOS Simulator 9.0 make iPhoneProxy hang if symlinks are used in the app bundle.
Using hardlinks instead fixes the issue, and the application can be started after deployment

peter-b commented:
@livecodesebastien You need to check the result of attempting to hardlink the files.  If they are on different filesystems (e.g. if the input file is on a USB stick and the output file is in a temp directory on the hard disk), hardlinking will fail, and you should fall back to copying.

@livecode-vulcan review not ok 9c361b3

livecode-vulcan commented:
💔 review by @peter-b not ok 9c361b38294fb6415de7bf8192ae336af799c5cb

livecodesebastien committed:
[[ Bug 16120 ]] Backup to copying if hardlinking fails

Cross-device hard-linking is not possible (for instance), so we simply copy the file if we can't hardlink it

livecodesebastien committed:
[[ Bug 16120 ]] Make a single shell command for hardlinking/copying

peter-b commented:
@livecode-vulcan review ok dc2af54

I think we finally got there. Thank you :100:

livecode-vulcan commented:
💙 review by @peter-b ok dc2af54f438dd2412b6a5f3602c06d9f82fa6ae6

livecode-vulcan commented:
😎 test success dc2af54f438dd2412b6a5f3602c06d9f82fa6ae6
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/719)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/732)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/720)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/722)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/732)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/743)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/734)

montegoulding commented:
@livecodesebastien I've had this report against my externals and LC 6.7.8 RC1 and I think it's related to this change:

I installed LC 6.7.8 (RC1) today, which supports iOS 9. I am using Mac OS 10.11 and xCode 7.0.1.

An iPad app that runs fine under iOS 8.x and xCode 6 now crashes instantly on startup. I don’t see any part of the app’s UI and it appears that its PreOpenStack handler is never called.

The app uses the mergDoc, mergPop and mergSettings externals. I am using the recently updated versions of all three externals. If I remove these three externals the app launches to its authentication screen, at which point it hangs as mergSettings is not available. Leaving any of the externals in the app causes an instant crash of the app. Removing all 3 allows the app to launch.

The console reports (below): Startup error: failed to load external

The exception type is EXC_CRASH (SIGABRT)

```
Oct 15 12:43:46 Humble-Pulse Rakonto[9399]: Startup error: failed to load external
Oct 15 12:43:46 Humble-Pulse com.apple.CoreSimulator.SimDevice.0DB7395A-8AB2-42B5-BC14-9A9E685EB26F.launchd_sim[8366] (UIKitApplication:com.rakonto.app[0x58dc][9399]): Service exited due to signal: Abort trap: 6
Oct 15 12:43:46 Humble-Pulse SpringBoard[8383]: BSXPCMessage received error for message: Connection invalid
Oct 15 12:43:46 Humble-Pulse SpringBoard[8383]: UNNotificationRegistrarConnectionListener connection invalidated
Oct 15 12:43:46 Humble-Pulse SpringBoard[8383]: BSXPCMessage received error for message: Connection invalid
Oct 15 12:43:46 Humble-Pulse assertiond[8387]: assertion failed: 15A284 13A340: assertiond + 13207 [28CC4371-F3F9-3578-9436-310B23A2C638]: 0x1
Oct 15 12:43:46 Humble-Pulse SpringBoard[8383]: Application 'UIKitApplication:com.rakonto.app[0x58dc]' crashed.
Oct 15 12:43:46 Humble-Pulse assertiond[8387]: assertion failed: 15A284 13A340: assertiond + 13207 [28CC4371-F3F9-3578-9436-310B23A2C638]: 0x1
Oct 15 12:43:46 Humble-Pulse assertiond[8387]: notify_suspend_pid() failed with error 7
Oct 15 12:43:46 Humble-Pulse assertiond[8387]: assertion failed: 15A284 13A340: assertiond + 13207 [28CC4371-F3F9-3578-9436-310B23A2C638]: 0x1
```

montegoulding commented:
After checking the code I see that externals aren't symlinked so it myst be something else.

montegoulding commented:
I actually did work out that the above issue is a regression on this PR. The problem is lcext files have been added to the redirects when they shouldn't have been. This wasn't an issue until the introduction of copy on symlink failure because the symlink was failing as a result of the the destination file existing. So there's two solutions: 
1. Ensure lcext files aren't added to the redirects (you should also handle other special cases like Info.plist and fonts)
2. exit if there is a file pTarget in symlinkRedirect
